### ISSUE TITLE
`calico/node` stuck after `daemon.go 658: Felix is shutting down reason="config changed"`

### ISSUE THREAD

**Issue by seeker89** [projectcalico/felix#2547]
TL;DR: When adding new nodes to a k8s cluster, the existing `calico/node` instances attempt to restart and seemingly get stuck.

## Expected Behavior
The old `calico-node` pods continue working when adding new `kind: Node` configuration.

## Current Behavior
When I introduce new nodes to an existing cluster, pod-to-pod networking breaks.

All of the previous `calico/node` pods print `Felix is shutting down reason="config changed"` like below, but don't actually restart, and stop working:

```
2020-10-22 14:03:52.560 [WARNING][45] daemon.go 1029: Config change requires restart key="IPv4VXLANTunnelAddr" old="10.244.82.0" updateType="delete"
2020-10-22 14:03:52.560 [WARNING][45] daemon.go 658: Felix is shutting down reason="config changed"
2020-10-22 14:03:54.629 [WARNING][8217] int_dataplane.go 404: Can't enable XDP acceleration. error=kernel is too old (have: 4.15.0 but want at least: 4.16.0)
```

I'm not familiar with the code base, but from the logs, I can conclude that it hits this message:

https://github.com/projectcalico/felix/blob/655eb3e5d40bece4f8f2d584e5dcacdb9c2fc607/daemon/daemon.go#L699-L700

But then it appears to be stuck in this wait:

https://github.com/projectcalico/felix/blob/655eb3e5d40bece4f8f2d584e5dcacdb9c2fc607/daemon/daemon.go#L704-L713

The process never exits. As far as docker is concerned, it's still running:

```sh
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 20979,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2020-10-21T17:11:45.960040614Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
```

I also looked at the master branch, that part of the code doesn't seem to have changed since the version `v3.13.3` that I'm running.

## Possible Solution
Walkaround: restarting `calico-node` containers brings the pod-to-pod networking back online.

## Steps to Reproduce (for bugs)
1. Create a Kubernetes cluster with N worker nodes, provision calico as the CNI using the `calico/node` container image as a DaemonSet
2. The cluster works fine
3. Provision extra worker nodes the same way
4. Create the extra `projectcalico.org/v3` `Node` yaml and apply
5. Watch the new `calico-nodes` pods come up OK on the new nodes, and the old ones get stuck in 'shutting down' on the previous nodes. Pod-to-pod networking no longer works.

## Context
This used to work before, but now I can reliably break the pod-to-pod networking by adding new workers to the cluster.

## Your Environment
* `calico/node` container version (if running Felix in a container): `calico/node:v3.13.3`, `calico/cni:v3.13.3`, `calico/kube-controllers:v3.13.3`
* Which orchestrator are you using (e.g. Kubernetes, OpenStack, Docker, Mesos): Kubernetes
* Orchestrator version (if used): v15.5
* Etcd version (if using etcd datastore driver):  `etcd:v3.3.15`
* Operating System and version: `Ubuntu 18.04.4 LTS`

Let me know if I can add any more useful details. Thanks!

**Comment by fasaxc** [comment]
@seeker89 I think I see a different problem in your log.  Felix's logs all include the PID of the Felix process (within the square brackets after the log level).  The fact that the PID changes indicates a successful restart.  I'm assuming that you've set the log level to warning, rather than the default info level, so you get no output from the restart unless it goes wrong.

However, the log about deletion of the VXLAN tunnel address is suspicious. I don't think adding a node should cause _other_ nodes to delete their VXLAN tunnel address.

When you add a node to the system, does your provisioning system also update the existing nodes?

**Comment by erhudy** [comment]
I've been working with @seeker89 on this issue and I have a reliable reproducer. When I change the Calico configuration slightly, Felix does indeed restart, but it doesn't repopulate VXLAN routes until the pod itself is deleted:
```
2020-10-28 15:22:51.133 [DEBUG][219] int_dataplane.go 1229: couldn't complete deferred work for manager, will try again later error=no encap route table not set, will defer adding routes manager=&intdataplane.vxlanManager{Mutex:sync.Mutex{state:0, sema:0x0}, hostname:"nedtest-master-1", routeTable:(*routetable.RouteTable)(0xc0001360f0), noEncapRouteTable:intdataplane.routeTable(nil), routesByDest:map[string]*proto.RouteUpdate{}, vtepsByNode:map[string]*proto.VXLANTunnelEndpointUpdate{}, myVTEP:(*proto.VXLANTunnelEndpointUpdate)(nil), vxlanDevice:"vxlan.calico", vxlanID:4096, vxlanPort:4789, routesDirty:true, ipsetsDataplane:(*ipsets.IPSets)(0xc00015be60), ipSetMetadata:ipsets.IPSetMetadata{SetID:"all-vxlan-net", Type:"hash:net", MaxSize:1048576}, externalNodeCIDRs:[]string(nil), vtepsDirty:false, nlHandle:(*netlink.Handle)(0xc00026f390), dpConfig:intdataplane.Config{Hostname:"nedtest-master-1", IPv6Enabled:false, RuleRendererOverride:rules.RuleRenderer(nil), IPIPMTU:1440, VXLANMTU:1410, MaxIPSetSize:1048576, IptablesBackend:"legacy", IPSetsRefreshInterval:10000000000, RouteRefreshInterval:90000000000, DeviceRouteSourceAddress:net.IP(nil), DeviceRouteProtocol:3, RemoveExternalRoutes:true, IptablesRefreshInterval:90000000000, IptablesPostWriteCheckInterval:1000000000, IptablesInsertMode:"insert", IptablesLockFilePath:"/run/xtables.lock", IptablesLockTimeout:0, IptablesLockProbeInterval:50000000, XDPRefreshInterval:90000000000, NetlinkTimeout:10000000000, RulesConfig:rules.Config{IPSetConfigV4:(*ipsets.IPVersionConfig)(0xc000130280), IPSetConfigV6:(*ipsets.IPVersionConfig)(0xc0001304b0), WorkloadIfacePrefixes:[]string{"cali"}, IptablesMarkAccept:0x10000, IptablesMarkPass:0x20000, IptablesMarkScratch0:0x40000, IptablesMarkScratch1:0x80000, IptablesMarkEndpoint:0xfff00000, IptablesMarkNonCaliEndpoint:0x100000, KubeNodePortRanges:[]numorstring.Port{numorstring.Port{MinPort:0x7530, MaxPort:0x7fff, PortName:""}}, KubeIPVSSupportEnabled:false, OpenStackMetadataIP:net.IP(nil), OpenStackMetadataPort:0x2247, OpenStackSpecialCasesEnabled:false, VXLANEnabled:true, VXLANPort:4789, VXLANVNI:4096, IPIPEnabled:false, IPIPTunnelAddress:net.IP(nil), VXLANTunnelAddress:net.IP(nil), IptablesLogPrefix:"calico-packet", EndpointToHostAction:"ACCEPT", IptablesFilterAllowAction:"ACCEPT", IptablesMangleAllowAction:"ACCEPT", FailsafeInboundHostPorts:[]config.ProtoPort{config.ProtoPort{Protocol:"tcp", Port:0x16}, config.ProtoPort{Protocol:"udp", Port:0x44}, config.ProtoPort{Protocol:"tcp", Port:0xb3}, config.ProtoPort{Protocol:"tcp", Port:0x94b}, config.ProtoPort{Protocol:"tcp", Port:0x94c}, config.ProtoPort{Protocol:"tcp", Port:0x1a0a}, config.ProtoPort{Protocol:"tcp", Port:0x1a0b}}, FailsafeOutboundHostPorts:[]config.ProtoPort{config.ProtoPort{Protocol:"udp", Port:0x35}, config.ProtoPort{Protocol:"udp", Port:0x43}, config.ProtoPort{Protocol:"tcp", Port:0xb3}, config.ProtoPort{Protocol:"tcp", Port:0x94b}, config.ProtoPort{Protocol:"tcp", Port:0x94c}, config.ProtoPort{Protocol:"tcp", Port:0x1a0a}, config.ProtoPort{Protocol:"tcp", Port:0x1a0b}}, DisableConntrackInvalid:false, NATPortRange:numorstring.Port{MinPort:0x0, MaxPort:0x0, PortName:""}, IptablesNATOutgoingInterfaceFilter:"", NATOutgoingAddress:net.IP(nil), BPFEnabled:false}, IfaceMonitorConfig:ifacemonitor.Config{InterfaceExcludes:[]*regexp.Regexp{(*regexp.Regexp)(0xc0000c54a0)}}, StatusReportingInterval:0, ConfigChangedRestartCallback:(func())(0x189a450), PostInSyncCallback:(func())(0x187ef90), HealthAggregator:(*health.HealthAggregator)(0xc0002fb920), DebugSimulateDataplaneHangAfter:0, ExternalNodesCidrs:[]string(nil), BPFEnabled:false, BPFKubeProxyIptablesCleanupEnabled:true, BPFLogLevel:"off", BPFDataIfacePattern:(*regexp.Regexp)(0xc0000c4fa0), XDPEnabled:true, XDPAllowGeneric:false, BPFConntrackTimeouts:conntrack.Timeouts{CreationGracePeriod:10000000000, TCPPreEstablished:20000000000, TCPEstablished:3600000000000, TCPFinsSeen:30000000000, TCPResetSeen:40000000000, UDPLastSeen:60000000000, ICMPLastSeen:5000000000}, BPFCgroupV2:"", BPFConnTimeLBEnabled:true, BPFMapRepin:true, BPFNodePortDSREnabled:false, KubeProxyMinSyncPeriod:1000000000, SidecarAccelerationEnabled:false, LookPathOverride:(func(string) (string, error))(nil), KubeClientSet:(*kubernetes.Clientset)(0xc0000d9b80)}, noEncapProtocol:80, noEncapRTConstruct:(func([]string, uint8, bool, time.Duration, net.IP, int, bool) intdataplane.routeTable)(0x1868e50)}
```

**Comment by fasaxc** [comment]
That log appears to indicate that felix is missing some VXLAN configuration that should be in the datastore.  Can you try dumping the Calico Node resource with `calicoctl`? 
```
calicoctl get node <name> -o yaml
```
should do the trick?

This log: 
```
2020-10-22 14:03:52.560 [WARNING][45] daemon.go 1029: Config change requires restart key="IPv4VXLANTunnelAddr" old="10.244.82.0" updateType="delete"
```
makes me think that some component is deleting the VXLAN tunnel address, which is causing felix to fail to program routes.  Restarting calico/node will trigger the VXLAN IP to be recalculated but it shouldn't be getting deleted at all.

**Comment by erhudy** [comment]
Ah, interesting. We try to keep Calico's configuration relatively idempotent by writing the whole thing to disk as `calico-configuration.yaml` and then doing `calicoctl apply -f` on that file. In my tiny test cluster, this YAML file on disk looks like this:
```
---
apiVersion: projectcalico.org/v3
kind: Node
metadata:
  name: nedtest-master-1
spec:
  bgp:
    ipv4Address: 10.42.1.58/18
---
apiVersion: projectcalico.org/v3
kind: Node
metadata:
  name: nedtest-worker-1
spec:
  bgp:
    ipv4Address: 10.42.2.8/18
---
apiVersion: projectcalico.org/v3
kind: IPPool
metadata:
  name: 10-244-0-0-16
spec:
  cidr: 10.244.0.0/16
  ipipMode: Never
  vxlanMode: Always
  natOutgoing: true
  blockSize: 26
  disabled: False
```
My guess is that when this file is applied, it's not merging the Node objects in the right way. The nodes usually look like this:
```
apiVersion: projectcalico.org/v3
items:
- apiVersion: projectcalico.org/v3
  kind: Node
  metadata:
    annotations:
      projectcalico.org/kube-labels: '{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/instance-type":"0b9965cb-439a-47f5-8426-3220ea8ee621","beta.kubernetes.io/os":"linux","failure-domain.beta.kubernetes.io/region":"dev","failure-domain.beta.kubernetes.io/zone":"dev-3","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"nedtest-master-1","kubernetes.io/os":"linux","node-role.kubernetes.io/master":""}'
    creationTimestamp: "2020-10-27T18:45:44Z"
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/instance-type: 0b9965cb-439a-47f5-8426-3220ea8ee621
      beta.kubernetes.io/os: linux
      failure-domain.beta.kubernetes.io/region: dev
      failure-domain.beta.kubernetes.io/zone: dev-3
      kubernetes.io/arch: amd64
      kubernetes.io/hostname: nedtest-master-1
      kubernetes.io/os: linux
      node-role.kubernetes.io/master: ""
    name: nedtest-master-1
    resourceVersion: "220326"
    uid: 96fb3767-c800-41c2-aacb-b5a9b2504754
  spec:
    bgp:
      ipv4Address: 10.42.1.58/32
    ipv4VXLANTunnelAddr: 10.244.193.134
    orchRefs:
    - nodeName: nedtest-master-1
      orchestrator: k8s
- apiVersion: projectcalico.org/v3
  kind: Node
  metadata:
    annotations:
      projectcalico.org/kube-labels: '{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/instance-type":"0b9965cb-439a-47f5-8426-3220ea8ee621","beta.kubernetes.io/os":"linux","failure-domain.beta.kubernetes.io/region":"dev","failure-domain.beta.kubernetes.io/zone":"dev-1","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"nedtest-worker-1","kubernetes.io/os":"linux","node-role.kubernetes.io/worker":""}'
    creationTimestamp: "2020-10-27T18:45:44Z"
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/instance-type: 0b9965cb-439a-47f5-8426-3220ea8ee621
      beta.kubernetes.io/os: linux
      failure-domain.beta.kubernetes.io/region: dev
      failure-domain.beta.kubernetes.io/zone: dev-1
      kubernetes.io/arch: amd64
      kubernetes.io/hostname: nedtest-worker-1
      kubernetes.io/os: linux
      node-role.kubernetes.io/worker: ""
    name: nedtest-worker-1
    resourceVersion: "220338"
    uid: 76709f47-981d-4583-8feb-e459368efc61
  spec:
    bgp:
      ipv4Address: 10.42.2.8/32
    ipv4VXLANTunnelAddr: 10.244.32.217
    orchRefs:
    - nodeName: nedtest-worker-1
      orchestrator: k8s
kind: NodeList
metadata:
  resourceVersion: "229898"
```

...but if I add some arbitrary label to the nodes in my on-disk file and then reapply that, the VXLAN tunnel information gets obliterated.

**Comment by erhudy** [comment]
I suspect that I might be able to work around this by just not writing the Node objects to disk for VXLAN, because there's nothing relevant in there (we don't need the BGP stuff for example) and they should be populated via the Calico Kubernetes controller.

**Comment by fasaxc** [comment]
Ah yes, our `apply` function doesn't merge resources; we do have a `patch` command in recent versions that does that.  We were intending to match `kubectl`'s semantics when we wrote `calicoct`l but I think the nuance that `kubectl` does a merge on apply wasn't picked up.

**Comment by erhudy** [comment]
Yeah, this actually explains a lot, I think we'd been operating under the impression that calicoctl's `apply` acted like `kubectl` and didn't read the help text that actually says the resource gets replaced.

**Comment by fasaxc** [comment]
Sorry about the confusion!

**Comment by seeker89** [comment]
Thanks for your help tracking this down @fasaxc and @erhudy!
